batch/src/main/java/org/openmrs/analytics/FhirEtl.java

```diff
@@ -148,55 +168,95 @@ static OpenmrsUtil createOpenmrsUtil(String sourceUrl, String sourceUser, String
+			fhirSearchUtil = new FhirSearchUtil(openmrsUtil);
+			parquetUtil = new ParquetUtil();
 		}
 		
 		@ProcessElement
-		public void ProcessElement(@Element SearchSegmentDescriptor segment, OutputReceiver<DomainResource> out) {
-			Bundle pageBundle = fhirSearchUtil.searchByPage(segment.pageId(), segment.count(), segment.pageOffset(),
-			    SummaryEnum.DATA);
-			fhirSearchUtil.uploadBundleToCloud(pageBundle);
+		public void ProcessElement(@Element SearchSegmentDescriptor segment, OutputReceiver<GenericRecord> out) {
+			Bundle pageBundle = fhirSearchUtil.searchByUrl(segment.searchUrl(), segment.count(), SummaryEnum.DATA);
+			if (parquetFile.isEmpty()) {
+				fhirStoreUtil.uploadBundleToCloud(pageBundle, fhirContext);
+			} else {
+				for (GenericRecord record : parquetUtil.generateRecords(pageBundle, fhirContext)) {
+					out.output(record);
+				}
+			}
+		}
+	}
+	
+	static String findSearchedResource(String search) {
+		int argsStart = search.indexOf('?');
+		if (argsStart >= 0) {
+			return search.substring(0, argsStart);
 		}
+		return search;
 	}
 	
 	static void runFhirFetch(FhirEtlOptions options, FhirContext fhirContext) throws CannotProvideCoderException {
-		List<SearchSegmentDescriptor> segments = createSegments(options, fhirContext);
-		if (segments.isEmpty()) {
+		ParquetUtil parquetUtil = new ParquetUtil();
+		Map<String, List<SearchSegmentDescriptor>> segmentMap = createSegments(options, fhirContext);
+		if (segmentMap.isEmpty()) {
 			return;
 		}
 		
 		Pipeline p = Pipeline.create(options);
-		p.apply(Create.of(segments)).apply(ParDo.of(new FetchSearchPageFn(options.getFhirStoreUrl(),
-		        options.getServerUrl() + options.getServerFhirEndpoint(), options.getUsername(), options.getPassword())));
+		for (Map.Entry<String, List<SearchSegmentDescriptor>> entry : segmentMap.entrySet()) {
+			String outputFile = "";
+			if (!options.getOutputParquetBase().isEmpty()) {
+				outputFile = options.getOutputParquetBase() + entry.getKey();
+			}
+			String resourceType = findSearchedResource(entry.getKey());
+			Schema schema = parquetUtil.getResourceSchema(resourceType, fhirContext);
+			PCollection<SearchSegmentDescriptor> inputSegments = p.apply(Create.of(entry.getValue()));
+			PCollection<GenericRecord> records = inputSegments
+			        .apply(ParDo.of(new FetchSearchPageFn(options.getGcpFhirStore(),
+			                options.getServerUrl() + options.getServerFhirEndpoint(), options.getOutputParquetBase(),
+			                options.getUsername(), options.getPassword())))
+			        .setCoder(AvroCoder.of(GenericRecord.class, schema));
+			ParquetIO.Sink sink = ParquetIO.sink(schema);
+			records.apply(FileIO.<GenericRecord> write().via(sink).to(outputFile));
+		}
 		p.run().waitUntilFinish();
 	}
 	
 	public static void main(String[] args) throws CannotProvideCoderException {
 		// Todo: Autowire
-		FhirContext fhirContext = FhirContext.forR4();
+		FhirContext fhirContext = FhirContext.forDstu3();
```

Comment:
I think for greatest flexibility we should allow each subproject to use both R3(DSTU3) and R4, since the sink repo might only support a specific version of FHIR

Reply:
@kimaina I should have switched that in all subprojects, please let me know if I have missed anything.

@pmanko I agree that we should support both but I think that change is beyond this PR, hence I created #33 for now.

Reply:
agreed and thanks for creating the tickets